lib/wallaroo/core/common/state_step_creator.pony

```diff
@@ -0,0 +1,286 @@
+    _info(state_name)?(key)?
+
+  fun ref add(state_name: String, key: Key, info: Info^) =>
+    try
+      _info.insert_if_absent(state_name, Map[Key, Info])?(key) = info
+    end
+
+  fun contains(state_name: String, key: Key): Bool =>
+    try
+      apply(state_name, key)?
+      true
+    else
+      false
+    end
+
+  fun clone(): KeyToStepInfo[Info] iso^ =>
+    let c = recover iso KeyToStepInfo[Info].create() end
+
+    for (sn, k_i) in _info.pairs() do
+      for (key, info) in k_i.pairs() do
+        c.add(sn, key, info)
+      end
+    end
+
+    c
+
+  fun pairs(): Iter[(String, String, Info)] =>
+    """
+    Return an iterator over tuples where the first two values are the state name
+    and the key, and the last value is the info value.
+    """
+    Iter[(String, Map[String, Info] box)](_info.pairs()).
+      flat_map[(String, (String, Info))](
+        { (k_m) => Iter[String].repeat_value(k_m._1)
+          .zip[(String, Info)](k_m._2.pairs()) }).
+      map[(String, String, Info)](
+        { (x) => (x._1, x._2._1, x._2._2) })
+
+class KeyToStepInfoTag[Info: Any tag]
+  let _info: Map[String, Map[Key, Info]]
+
+  new create() =>
+    _info = _info.create()
+
+  fun apply(state_name: String, key: box->Key!): this->Info ? =>
+    _info(state_name)?(key)?
+
+  fun ref add(state_name: String, key: Key, info: Info^) =>
+    try
+      _info.insert_if_absent(state_name, Map[Key, Info])?(key) = info
+    end
+
+  fun contains(state_name: String, key: Key): Bool =>
+    try
+      apply(state_name, key)?
+      true
+    else
+      false
+    end
+
+  fun clone(): KeyToStepInfoTag[Info] iso^ =>
+    let c = recover iso KeyToStepInfoTag[Info].create() end
+
+    for (sn, k_i) in _info.pairs() do
+      for (key, info) in k_i.pairs() do
+        c.add(sn, key, info)
+      end
+    end
+
+    c
+
+  fun pairs(): Iter[(String, String, Info)] =>
+    """
+    Return an iterator over tuples where the first two values are the state name
+    and the key, and the last value is the info value.
+    """
+    Iter[(String, Map[String, Info] box)](_info.pairs()).
+      flat_map[(String, (String, Info))](
+        { (k_m) => Iter[String].repeat_value(k_m._1)
+          .zip[(String, Info)](k_m._2.pairs()) }).
+      map[(String, String, Info)](
+        { (x) => (x._1, x._2._1, x._2._2) })
+
+actor StateStepCreator is Initializable
+  var _keys_to_steps: KeyToStepInfoTag[Step] = _keys_to_steps.create()
+  var _keys_to_step_ids: KeyToStepInfo[StepId] = _keys_to_step_ids.create()
+
+  let _step_id_gen: StepIdGenerator = _step_id_gen.create()
+
+  let _app_name: String
+  let _worker_name: String
+  let _metrics_conn: MetricsSink
+  let _auth: AmbientAuth
+  let _event_log: EventLog
+  var _recovery_replayer: (None | RecoveryReplayer) = None
+
+  var _outgoing_boundaries: Map[String, OutgoingBoundary] val =
+    recover _outgoing_boundaries.create() end
+
+  var _runner_builders: Map[String, RunnerBuilder] val =
```

Comment:
Comment on what this is

Reply:
Suggestion (I'll be doing this on my branch as well): Create `state_name.pony` in `core/common`, create `type StateName is String` and let's start replacing these instances with `StateName` to make the intention clearer.

Reply:
Changed name to `_state_runner_builders`, so no need to comment. Created `StateName` and used it here, but not anywhere else.

Created #2261 to start using `StateName`.